### ISSUE TITLE
ED25519 keys

### ISSUE THREAD

**Issue by m00nwtchr** [tpm2-software/tpm2-pkcs11#785]
Would it be possible to add support for them, or is there a limitation which prevents that?

**Comment by williamcroberts** [comment]
I'm not really sure what ED25519, but a quick look into it, the TPM doesn't support it. Do you have or know of TPMs that support this?

**Comment by m00nwtchr** [comment]
Basically it's a ssh key type that uses an alternative curve (Curve25519) to the standard NIST P-256 one. I don't know any (though I don't really know where to find that information), but good to know that it's something that has to be specifically supported in the hardware itself.

**Comment by williamcroberts** [comment]
So TCG does know about it's in the registered algorithms https://trustedcomputinggroup.org/resource/tcg-algorithm-registry/ as TPM_ECC_CURVE_25519 but I don't find any mentions of it in the specification for the TPM https://trustedcomputinggroup.org/resource/tpm-library-specification/ and the PC Client Specification of https://trustedcomputinggroup.org/wp-content/uploads/TCG_PC_Client_Platform_TPM_Profile_PTP_Specification_Family_2.0_Revision_1.3v22.pdf. So my guess is their is no TPM that supports this in the wild and nothing in the pipeline.

**Comment by megla-tlanghorst** [comment]
Just FYI, it seems to be in the spec now, though not sure whether it's already supported anywhere: https://trustedcomputinggroup.org/wp-content/uploads/Trusted-Platform-Module-2.0-Library-Part-2-Version-184_pub.pdf#subsection.6.4
Though it does not seem to be in the PC recommendations (yet?): https://trustedcomputinggroup.org/wp-content/uploads/PC-Client-Specific-Platform-TPM-Profile-for-TPM-2p0-Version-1p06_pub.pdf

**Comment by makinikm** [comment]
If this gets implemented, it could significantly improve the open-source opsec generally – all the modern [software](https://en.wikipedia.org/wiki/EdDSA#Software) using Ed25519 SSH keys will inherently be able to use TPM-backed keys, securing these against exfiltration. Since EdDSA was accepted as standard, some projects have restricted use of older schemes due to known issues, this will expand the ecosystem secure key storage options, on x86 specifically.
(_Advocating out of personal interest in a project that shall remain nameless here and not listed on the wiki page either_)
:rocket:

**Comment by williamcroberts** [comment]
> If this gets implemented, it could significantly improve the open-source opsec generally – all the modern [software](https://en.wikipedia.org/wiki/EdDSA#Software) using Ed25519 SSH keys will inherently be able to use TPM-backed keys, securing these against exfiltration. Since EdDSA was accepted as standard, some projects have restricted use of older schemes due to known issues, this will expand the ecosystem secure key storage options, on x86 specifically. (_Advocating out of personal interest in a project that shall remain nameless here and not listed on the wiki page either_) 🚀

Do you know what hardware and/or simulators support this by any chance?

**Comment by makinikm** [comment]
> 
> Do you know what hardware and/or simulators support this by any chance?

To my best understanding, `"If your computer is based on the 8th Generation or later Intel® Core™ Processor family, then your system has Intel® PTT, an integrated TPM that adheres to the 2.0 specifications."` - from [here](https://www.intel.com/content/www/us/en/support/articles/000094205/processors/intel-core-processors.html)...

**Comment by williamcroberts** [comment]
> > Do you know what hardware and/or simulators support this by any chance?
> 
> To my best understanding, `"If your computer is based on the 8th Generation or later Intel® Core™ Processor family, then your system has Intel® PTT, an integrated TPM that adheres to the 2.0 specifications."` - from [here](https://www.intel.com/content/www/us/en/support/articles/000094205/processors/intel-core-processors.html)...

Yeah, just because the spec supports it, doesn't mean it's available in the wild yet. Also, just because it's in the spec, if it's marked optional in the pc client specification, tpm vendors still may omit it. So we would need to see if any intel machines have it, or if one of the emulators have it. We can't add a feature without a method to test it.

**Comment by makinikm** [comment]
Here's a simplified and personally tested [method](https://jade.fyi/blog/tpm-ssh/) of generating and using the keys from `tpm2-pkcs11`. The blog author devised it on an AMD system, and that exact method worked on a 7th generation Intel Core i5-7260U dual-core CPU (Nuc kit), which surprisingly has PTT despite that support article stating that it's present since gen 8.
Although no way to verify this, but imho - if you own a working Intel/AMD machine, you don't need any special hardware or simulators for testing this. Just make sure your TPM is enabled in firmware.

**Comment by williamcroberts** [comment]
> Here's a simplified and personally tested [method](https://jade.fyi/blog/tpm-ssh/) of generating and using the keys from `tpm2-pkcs11`. The blog author devised it on an AMD system, and that exact method worked on a 7th generation Intel Core i5-7260U dual-core CPU (Nuc kit), which surprisingly has PTT despite that support article stating that it's present since gen 8. 

That blog doesn't use ED25519 keys

> Although no way to verify this, but imho - if you own a working Intel/AMD machine, you don't need any special hardware or simulators for testing this. Just make sure your TPM is enabled in firmware.

Yes I have both and didn't see the curve in `tpm2_getcap ecc-curves` output (even if it cannot convert the name it'll output hex, so the tool should output it even if the algorithm is one it doesn't know about yet). So we need something that has support still. I see the constant defined in Stefan Berger's libtpms which is part of our testing framework. So we need to grab that and check it's tpm2_getcap output.

**Comment by makinikm** [comment]
> 
> That blog doesn't use ED25519 keys

Exactly! I mean, why not??
 
> 
> Yes I have both and didn't see the curve in `tpm2_getcap ecc-curves` output...

On that ancient Nuc machine I mentioned, the `tpm2_getcap ecc-curves`produced just this:
```
TPM2_ECC_NIST_P256: 0x3
TPM2_ECC_BN_P256: 0x10
```
And the `tpm2_getcap algorithms` amongst others printed these:
```
ecdsa:
  value:      0x18
  asymmetric: 1
  symmetric:  0
  hash:       0
  object:     0
  reserved:   0x0
  signing:    1
  encrypting: 0
  method:     1
ecdh:
  value:      0x19
  asymmetric: 1
  symmetric:  0
  hash:       0
  object:     0
  reserved:   0x0
  signing:    0
  encrypting: 0
  method:     1
ecdaa:
  value:      0x1A
  asymmetric: 1
  symmetric:  0
  hash:       0
  object:     0
  reserved:   0x0
  signing:    1
  encrypting: 0
  method:     0
ecschnorr:
  value:      0x1C
  asymmetric: 1
  symmetric:  0
  hash:       0
  object:     0
  reserved:   0x0
  signing:    1
  encrypting: 0
  method:     0
```
